biscuit/src/Auth/Biscuit/Datalog/Parser.hs

```diff
@@ -1,448 +1,393 @@
+  _ <- l $ chunk "()"
+  pure $ EUnary method e1
 
-binary :: HasTopTermParsers 'InPredicate ctx
-       => Text
-       -> Binary
-       -> Expr.Operator Parser (Expression' ctx)
-binary name op = Expr.InfixL  (EBinary op <$ (skipSpace *> string name))
-
-hexBsParser :: Parser ByteString
-hexBsParser = do
-  void $ string "hex:"
-  either (fail . Text.unpack) pure . Hex.decodeBase16 . encodeUtf8 =<< takeWhile1 (inClass "0-9a-fA-F")
-
-litStringParser :: Parser Text
-litStringParser =
-  let regularChars = takeTill (inClass "\"\\")
-      escaped = choice
-        [ string "\\n" $> "\n"
-        , string "\\\"" $> "\""
-        , string "\\\\"  $> "\\"
-        ]
-      str = do
-        f <- regularChars
-        r <- optional (liftA2 (<>) escaped str)
-        pure $ f <> fold r
-   in char '"' *> str <* char '"'
+methodParser :: Parser (Expression' 'WithSlices)
+methodParser = binaryMethodParser <|> unaryMethodParser
 
-rfc3339DateParser :: Parser UTCTime
-rfc3339DateParser =
-  -- get all the chars until the end of the term
-  -- a term can be terminated by
-  --  - a space (before another delimiter)
-  --  - a comma (before another term)
-  --  - a closing paren (the end of a term list)
-  --  - a closing bracket (the end of a set)
-  let getDateInput = takeWhile1 (notInClass ", )];")
-      parseDate = parseTimeM False defaultTimeLocale "%FT%T%Q%EZ"
-   in parseDate . unpack =<< getDateInput
-
-termParser :: forall inSet pof ctx
-            . ( HasTermParsers inSet pof ctx
-              )
-           => Parser (Term' inSet pof ctx)
-termParser = skipSpace *> choice
-  [ Antiquote <$> ifPresent "slice" (Slice <$> (string "${" *> haskellVariableParser <* char '}'))
-  , Variable <$> ifPresent "var" variableNameParser
-  , TermSet <$> parseSet @inSet @ctx
-  , LBytes <$> hexBsParser
-  , LDate <$> rfc3339DateParser
-  , LInteger <$> signed decimal
-  , LString <$> litStringParser
-  , LBool <$> choice [ string "true"  $> True
-                     , string "false" $> False
-                     ]
+unaryParens :: Parser (Expression' 'WithSlices)
+unaryParens = do
+  _ <- l $ C.char '('
+  e <- l expressionParser
+  _ <- l $ C.char ')'
+  pure $ EUnary Parens e
+
+exprTerm :: Parser (Expression' 'WithSlices)
+exprTerm = choice
+  [ unaryParens
+  , EValue <$> predicateTermParser
   ]
 
--- | same as a predicate, but allows empty
--- | terms list
-ruleHeadParser :: HasTopTermParsers 'InPredicate ctx => Parser (Predicate' 'InPredicate ctx)
-ruleHeadParser = do
-  skipSpace
-  name <- predicateNameParser
-  skipSpace
-  terms <- parens (commaList0 termParser)
-  pure Predicate{name,terms}
-
-ruleBodyParser :: (HasParsers 'InPredicate evalCtx ctx)
-               => Parser ([Predicate' 'InPredicate ctx], [Expression' ctx], Set.Set (RuleScope' evalCtx ctx))
+ruleParser :: Parser (Rule' 'Repr 'WithSlices)
+ruleParser = do
+  rhead <- l predicateParser
+  _ <- l $ chunk "<-"
+  (body, expressions, scope) <- ruleBodyParser
```

Comment:
Does it imply no space allowed around the left arrow ?

Reply:
Or that's the job of `l` ?

Reply:
that's `l`'s job. it handles whitespace (including comments)